### PR TITLE
river: allow referencing zero values in Go maps

### PR DESCRIPTION
**Pull request by rfratto**
This fixes an issue where any reference to a zero value in a Go map would be treated as if the key didn't exist. This was due to a misunderstanding of the Go documentation that said "the zero Value" will be returned if the key does not exist in the map. This documentation referred to an invalid `reflect.Value` instead of the zero value of the map's value type.

This allows to simplify the logic when performing a map lookup to determine if the key was found or not, and permits expressions which reference a map key set to the zero value of the value type.

Opening in draft because it still needs some unit tests.

### CODE REVIEW

File: pkg/river/internal/value/value_test.go [hunk 1]
```diff
@@ -64,6 +100,10 @@ func (*pointerMarshaler) MarshalText() ([]byte, error) {
 	return []byte("Hello, world!"), nil
 }
 
+type Body struct {
+	Data pointerMarshaler `river:"data,attr"`
+}
```
Comment (rfratto) [inline pkg/river/internal/value/value_test.go:105]: @erikbaranowski Since this is being declared for the scope of the one test, and it doesn't have any methods, can we move it to the top of the inside of the function?

```go
func TestInterfacePointerReceiver() {
  type Body { ... } 

  ...
}

File: pkg/river/internal/value/value_test.go [hunk 2]
```diff
@@ -13,29 +13,85 @@ import (
...
+		expectKeyExists bool
+		expectKeyValue  value.Value
+		expectKeyType   value.Type
+	}{
+		{
+			name:           "Encode",
+			encodeTarget:   &pointerMarshaler{},
+			expectBodyType: value.TypeString,
+			expectBodyText: "Hello, world!",
+		},
```
Comment (rfratto) [inline pkg/river/internal/value/value_test.go:33]: Can we split this out into its own test function? It feels a bit weird that the rest of these tests are more about testing key lookups but this one isn't

Comment (rfratto) [inline pkg/river/internal/value/value_test.go:33]: @erikbaranowski Can we split this out into its own test function? It feels a bit weird that the rest of these tests are more about testing key lookups but this one isn't

File: pkg/river/internal/value/value_test.go [hunk 3]
```diff
@@ -9,33 +9,107 @@ import (
...
+		name         string
+		encodeTarget any
+		key          string
+
+		expectBodyType  value.Type
+		expectKeyExists bool
+		expectKeyValue  value.Value
+		expectKeyType   value.Type
+	}{
+		{
```
Comment (rfratto) [inline pkg/river/internal/value/value_test.go:29]: @erikbaranowski Can you add a test case here for getting a zero value from a River object?

For example, for `map[string]string{"data": ""}` the lookup should succeed and the resulting value should be the empty string. 

After that, I think this is good to merge.
